docs/design/features/byreflike-generics.md

```diff
@@ -0,0 +1,123 @@
+- `Span<Span<char>>` &ndash; Nested `Span<T>` types would be of benefit in the parsing result of strings.
+
+## Runtime impact
+
+Supporting ByRefLike type as Generic parameters will impact the following IL instructions:
+
+- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
+- `stsfld` / `ldsfld` &ndash; Type fields of a ByRefLike parameter cannot be marked `static`.
+- `newarr` / `stelem` / `ldelem` / `ldelema` &ndash; Arrays are not able to contain ByRefLike types.
+    - `newobj` &ndash; For multi-dimensional array construction.
+- `constrained.callvirt` &ndash; If this IL sequence resolves to a method implemented on `object` or default interface method, an error will occur during the attempt to box the instance.
+
+If any of the above instructions are attempted to be used with a ByRefLike type, the runtime will throw an `InvalidProgramException`. Sequences involving some of the above instructions are considered optimizations and represent cases that will remain valid regardless of a `T` being ByRefLike. See "Special IL Sequences" section below for details.
+
+The following instructions are already set up to support this feature since their behavior will fail as currently defined due to the inability to box a ByRefLike type.
+
+- `throw` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `unbox` / `unbox.any` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `isinst` &ndash; Will always place `null` on stack.
+- `castclass` &ndash; Will always throw `System.InvalidCastException`.
+
+## API Proposal
+
+Support for the following will be indicated by a new property. For .NET 7, the feature will be marked with `RequiresPreviewFeaturesAttribute` to indicate it is in [preview](https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md).
+
+```diff
+namespace System.Runtime.CompilerServices
+{
+    public static partial class RuntimeFeature
+    {
++        /// <summary>
++        /// Represents a runtime feature where byref-like types can be used in Generic parameters.
++        /// </summary>
++        public const string GenericsAcceptByRefLike = nameof(GenericsAcceptByRefLike);
+    }
+}
+```
+
+The compiler will need an indication for existing troublesome APIs where ByRefLike types will be permissable, but where the failure will be handled at runtime. An attribute will be created and added to these APIs.
+
+```csharp
+namespace System.Runtime.CompilerServices
+{
+    /// <summary>
+    /// Indicates to the compiler that constrain checks should be suppressed
+    /// and will instead be enforced at run-time.
+    /// </summary>
+    public sealed class SuppressConstraintChecksAttribute : Attribute
+    { }
+}
+```
+
+Troublesome APIs:
+- [`Span<T>`](https://docs.microsoft.com/dotnet/api/system.span-1)
+    - `public Span(T[]? array);`
```

Comment:
The type loader rejects arrays of byref like types very early today (look for `IDS_CLASSLOAD_BYREFLIKEARRAY`). It is not possible to create a type for e.g. `Span<int>[]`. 

This restriction would prevent troublesome methods from being inspected using reflection when the type is instantiated over byref type. Attempt to get the method argument types would throw.

I think we will need to allow creating these types (e.g. allow `typeof(Span<int>[])`) and delay the failure till attempt to allocate the array allocation to make the reflection behavior reasonable.

Reply:
I think we are going with the idea that using Reflection on the `Span<T>`/`ReadOnlySpan<T>` types may cause a `TypeLoadException` if `T` is ByRefLike.